node/pkg/tss/comm/server.go

```diff
@@ -0,0 +1,293 @@
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	tsscommv1 "github.com/certusone/wormhole/node/pkg/proto/tsscomm/v1"
+	"github.com/certusone/wormhole/node/pkg/tss"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+type connection struct {
+	cc     *grpc.ClientConn
+	stream tsscommv1.DirectLink_SendClient
+}
+
+type redialResponse struct {
+	name string
+	conn *connection
+}
+
+type server struct {
+	tsscommv1.UnimplementedDirectLinkServer
+	ctx        context.Context
+	logger     *zap.Logger
+	socketPath string
+
+	tssMessenger tss.ReliableMessenger
+
+	peers      []*tsscommv1.PartyId
+	peerToCert map[string]*x509.Certificate
+	// to ensure thread-safety without locks, only the sender is allowed to change this map.
+	connections   map[string]*connection
+	requestRedial chan string
+	redials       chan redialResponse
+}
+
+func (s *server) run() {
+	go s.dailer()
+	go s.sender()
+
+	for _, pid := range s.peers {
+		s.enqueueRedialRequest(pid.Id)
+	}
+}
+
+func (s *server) sender() {
+	connectionCheckTicker := time.NewTicker(connectionCheckTime)
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			for _, con := range s.connections {
+				con.cc.Close()
+			}
+
+			return
+
+		case o := <-s.tssMessenger.ProducedOutputMessages():
+			s.send(o)
+
+		case redial := <-s.redials:
+			if _, ok := s.connections[redial.name]; ok {
+				redial.conn.cc.Close() // shouldn't open the same connection twice.
+
+				continue
+			}
+
+			s.connections[redial.name] = redial.conn
+
+		case <-connectionCheckTicker.C:
+			// this case is an ensurance.
+			s.ensuredConnected()
+		}
+	}
+}
+
+func (s *server) ensuredConnected() {
+	if len(s.connections) != len(s.peers) {
+		for _, pid := range s.peers {
+			hostname := pid.Id
+			if _, ok := s.connections[hostname]; !ok {
+				s.enqueueRedialRequest(hostname)
+			}
+		}
+	}
+}
+
+func (s *server) send(msg *tsscommv1.PropagatedMessage) {
+	switch msg.Payload.(type) {
+	case *tsscommv1.PropagatedMessage_Echo:
+		s.broadcast(msg)
+
+	case *tsscommv1.PropagatedMessage_Unicast:
+		s.unicast(msg)
+	}
+}
+
+func (s *server) unicast(msg *tsscommv1.PropagatedMessage) {
+	m, ok := msg.Payload.(*tsscommv1.PropagatedMessage_Unicast)
+	if !ok {
+		return // shouldn't happen.
+	}
+
+	for _, recipient := range m.Unicast.Recipients {
+		hostname := recipient.Id
+
+		conn, ok := s.connections[recipient.Id]
+		if !ok {
+			s.enqueueRedialRequest(hostname)
+
+			s.logger.Error(
+				"received unknown recipient",
+				zap.String("hostname", recipient.Id),
+			)
+
+			continue
+		}
+
+		if err := conn.stream.Send(msg); err != nil {
+			delete(s.connections, hostname)
+			s.enqueueRedialRequest(hostname)
+
+			s.logger.Warn(
+				"couldn't send message to peer.",
+				zap.String("hostname", recipient.Id),
+				zap.Error(err),
+			)
+		}
+	}
+}
+
+func (s *server) broadcast(msg *tsscommv1.PropagatedMessage) {
+	for id, conn := range s.connections {
```

Comment:
A stream error leads to stream termination in the gRPC library. 
As a result, I deleted it from the `s.connections` map (Since it isn't usable anymore).
Eventually, the connection would resume (due to the dialer goroutine).

If the server can't reach at least `f+1` servers (`len(s.connections)< f+1`), then this server is one of the `f` faulty allowed by the reliable-broadcast protocol and won't be able to participate.